concordium-std/src/impls.rs

```diff
@@ -2373,48 +2402,37 @@ where
     }
 }
 
-#[cfg(test)]
-/// Some helper methods that are used for internal tests.
-impl<S> StateBuilder<S>
-where
-    S: HasStateApi,
-{
-    /// Get a value from the generic map.
-    /// `Some(Err(_))` means that something exists in the state with that key,
-    /// but it isn't of type `V`.
-    pub(crate) fn get<K: Serial, V: DeserialWithState<S>>(&self, key: K) -> Option<ParseResult<V>> {
-        let key_with_map_prefix = Self::prepend_generic_map_key(key);
-
-        self.state_api
-            .lookup_entry(&key_with_map_prefix)
-            .map(|mut entry| V::deserial_with_state(&self.state_api, &mut entry))
+impl StateBuilder<StateApi> {
+    /// Create a new empty [`StateBTreeSet`](crate::StateBTreeSet).
+    pub fn new_btree_set<K>(&mut self) -> state_btree::StateBTreeSet<K> {
+        let (state_api, prefix) = self.new_state_container();
+        state_btree::StateBTreeSet::new(state_api, prefix)
     }
 
-    /// Inserts a value in the generic map.
-    /// The key and value are serialized before insert.
-    pub(crate) fn insert<K: Serial, V: Serial>(
-        &mut self,
-        key: K,
-        value: V,
-    ) -> Result<(), StateError> {
-        let key_with_map_prefix = Self::prepend_generic_map_key(key);
-        match self.state_api.entry(key_with_map_prefix) {
-            EntryRaw::Vacant(vac) => {
-                let _ = vac.insert(&value);
-            }
-            EntryRaw::Occupied(mut occ) => occ.insert(&value),
+    /// Create a new empty [`StateBTreeMap`](crate::StateBTreeMap).
+    pub fn new_btree_map<K, V>(&mut self) -> state_btree::StateBTreeMap<K, V> {
+        state_btree::StateBTreeMap {
+            key_value: self.new_map(),
+            key_order: self.new_btree_set(),
         }
-        Ok(())
     }
 
-    /// Serializes the key and prepends [GENERIC_MAP_PREFIX].
-    /// This is similar to how [StateMap] works, where a unique prefix is
-    /// prepended onto keys. Since there is only one generic map, the prefix
-    /// is a constant.
-    fn prepend_generic_map_key<K: Serial>(key: K) -> Vec<u8> {
-        let mut key_with_map_prefix = to_bytes(&GENERIC_MAP_PREFIX);
-        key_with_map_prefix.append(&mut to_bytes(&key));
-        key_with_map_prefix
+    /// Create a new empty [`StateBTreeSet`](crate::StateBTreeSet), setting the
+    /// minimum degree `M` of the B-Tree explicitly.
+    pub fn new_btree_set_degree<const M: usize, K>(&mut self) -> state_btree::StateBTreeSet<K, M> {
+        let (state_api, prefix) = self.new_state_container();
+        state_btree::StateBTreeSet::new(state_api, prefix)
+    }
```

Comment:
I would prefer if we checked here that the degree is not invalid.

I think something like
```
    pub fn new_btree_set_degree<const M: usize, K>(&mut self) -> state_btree::StateBTreeSet<K, M> {
        if M >= 2 {
            let (state_api, prefix) = self.new_state_container();
            state_btree::StateBTreeSet::new(state_api, prefix)
        } else {
             // panic or maybe we can do report_error
        }
    }
```

in any smart contract code the M should be fully instantiated at some point so the if should be optimized away unless the person is using a wrong M.

Reply:
Yes, that seems like a good idea